### ISSUE TITLE
Cancel pending replies

### ISSUE THREAD

**Issue by sssoleileraaa** [freedomofpress/securedrop-client#810]
# Description

This issue is a place to capture thoughts on whether or not we should provide a cancel option for pending replies, and if we do, what that would look like in the GUI.

## What is a pending reply?

A pending reply is a reply that the client has in its queue to send off to the server. It is in a pending state until it succeeds or fails for any reason other than a read timeout or connect timeout issue. 

## What is a failed reply?

A failed reply is a reply that has either failed for any reason other than a timeout issue, or failed because it was in a pending state when the client was closed/ the user logged out.

## What if a pending reply continues to see timeout errors?

If a pending reply sees timeouts while being processed, the queue will continue to retry this job until the queue is paused, and the job will remain in the queue as the next job to process and in a pending state (because it has not succeeded or failed for any reason other than a read timeout or connect timeout issue).

And this brings up the question, **should there be a way to cancel a pending reply?**

Here are a few different ideas:

* A) If we don't have an expiration on how long we will keep a reply in a pending state, or if we don't provide a way to cancel a pending reply, then the only way a user can stop a reply from being sent will be to log out of the client (or close it). And, just in case someone wants to see an implementation of this, you can test this with the gmail phone app (when I sent messages on a slow network in gmail I noticed they were marked as "Queued" with no cancel option, and I assume there was no expiration but there could have been... also these queued messages might remain queued across sessions-- I didn't test this).

* B) If we want to provide a cancel option next to a reply, we will have to decide whether or not that means we remove it from the queue, or drop the reply once it's being processed (currently this is much simpler to implement because we don't have a way to remove specified jobs from the queue). If we wait until the job is processed before dropping it, there is an in-between state to handle. We could either update the GUI immediately to no longer show the cancelled pending reply or we could mark as a cancelled pending reply (a UX design for this will be needed). And for the backend we would need to make sure to delete pending replies that are marked as cancelled.

* C) Another idea to consider is an expiration on pending replies so that the client doesn't attempt to send a reply that was sent by the user X minutes/hours ago, and instead shows the reply as failed if it's passed the expiration date. This gives the user less control, unless it is implemented along with option B above.

* D) Provide a way to cancel  _all_ jobs in the queue (file downloads, sending replies, etc)

**Comment by ninavizz** [comment]
Time for that activity log, guys. I'd rather not halfass that feature... and either just do it, or not. I'm also sick and possibly a little grouchy, so happy to talk this through.

**Comment by creviera** [comment]
> > D) Provide a way to cancel all jobs in the queue (file downloads, sending replies, etc)
> 
> I currently favor that approach, precisely _because_ it doesn't give any guarantees about what jobs will be cancelled. Given that user actions may contend with repeated efforts to retry failing jobs, an imprecise "give up, cancel all" seems to me to me the most reasonable approach that avoids mysteriously disappearing "Cancel" links, race conditions, and the like.

I think this makes sense too, and it's more user-friendly than forcing the user to log out to cancel jobs. We also might see from user research that adding a way to cancel a pending reply and not everything else is desired as well.

**Comment by ninavizz** [comment]
Yeah, I really dislike "Cancel All." I'd rather we make product feature decisions around what is the best solution for users vs what's easiest to implement.

**Comment by ninavizz** [comment]
Also: leaving this here... https://invis.io/KXS6I4C4QA2#/365077763_v2_-_Pane_Open

**Comment by ninavizz** [comment]
![Active Queue](https://user-images.githubusercontent.com/8262612/74876910-6a3ea680-5319-11ea-8a82-cb9b88fc881e.png)
^ ...also, this was created at some point.

**Comment by zenmonkeykstop** [comment]
Related to https://github.com/freedomofpress/securedrop-client/pull/1632 - we need to revisit reply behaviour and consistency across clients/server in general.
